### ISSUE TITLE
Why does numpy fromfile need flush and fileno?

### ISSUE THREAD

**Issue by MaxNoe** [numpy/numpy#7831]
As for as I know, `fp.flush()` is only ever needed for write operations.

So why does `numpy.fromfile` need a `fp.flush`?

```{python}
import numpy as np


class MyFile:
    def read(self):
        return b'\x00\x01\x02\x03'

fp = MyFile()
print(np.fromfile(fp, 'u2'))
```

results in:

```
AttributeError: 'MyFile' object has no attribute 'flush'
```

Adding a  method like this:

```
def flush(self):
    pass
```

Results in 
```
TypeError: argument must be an int, or have a fileno() method.
```

What do I have to provide to make my own file-like object work? Why is a simple `read()` not enough?

**Comment by andyfaff** [comment]
Is there any reason you can't use StringIO?

**Comment by MaxNoe** [comment]
No. There is also no reason why i cannot use `numpy.frombuffer(f.read(), 'u2')`.

That's not the point. I simply do not expect a read-only operation to require a `flush` method or a `fileno` where a having a `read` (and maybe a `seek` if not the whole file should be read) should be sufficient.

This is unexpected and unpythonic.

**Comment by rkern** [comment]
`np.fromfile()` needs an actual `file` not a `file`-like object. It uses C routines that work on the underlying file descriptors, not the Python `.read()` API.

**Comment by pslacerda** [comment]
Using `flush` on a read operation continues being very unexpected.

**Comment by rkern** [comment]
@pv is probably better placed to answer this question, as he introduced this change. My reading of the source indicates that we actually do the file I/O on a `dup()`ed file descriptor. A `flush()` is required prior to the `dup()` to make sure that the file descriptor's state reflects the `FILE*` pointer, which may be buffered. `np.fromfile()` works on files opened as `'r+'`, not just read-only file objects.

**Comment by pslacerda** [comment]
If no process is forked and is a simple plain read why the `dup()`? Also a simple search for 'dup' or 'flush' returns no information relative to `np.fromfile()`, just distutils or debugging things.

**Comment by pv** [comment]
numpy.fromfile requires a C stdlib FILE* handle. The rest follows from that.

**Comment by pslacerda** [comment]
Of course with real file operations done at `fromfile_next_element()` and `fromfile_skip_separator()`, I just can't find the `PyArray_ArrFuncs->scanfunc` implementation.

**Comment by rkern** [comment]
@pslacerda [Here](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/multiarraymodule.c#L2060) is where the `dup()` happens.

[Here is the definition of `npy_PyFile_Dup2`](https://github.com/numpy/numpy/blob/master/numpy/core/include/numpy/npy_3kcompat.h#L148-L182) for Python 3. The `flush()` and `dup()` are only necessary in Python 3 because of the I/O overhaul there.

**Comment by pslacerda** [comment]
Very thank you!

There is any real need for `npy_PyFile_Check()` as it is unused? By the way, it has the same logic of the Python 2 `npy_PyFile_Dup2()`.

`dup2` is a system call name and must not be used as function name because it implies something else. Or maybe am I wrong?

From the [manual][1] `dup` has no obvious application use usage, so what it does here? I can't even get the package compiled without it =)

[1]: http://pubs.opengroup.org/onlinepubs/009695399/functions/dup.html

**Comment by pslacerda** [comment]
Of course it's used! What happened to me to say that?
